src/vset.h

```diff
@@ -0,0 +1,317 @@
+ *
+ * Expiry Retrieval/Removal:
+ *     long long vsetEstimatedEarliestExpiry(vset *set, vsetGetExpiryFunc getExpiry);
+ *     size_t vsetPopExpired(vset *set, vsetGetExpiryFunc getExpiry, vsetExpiryFunc expiryFunc, mstime_t now, size_t max_count, void *ctx);
+ *
+ * Utilities:
+ *     bool vsetIsEmpty(vset *set);
+ *
+ * Iteration:
+ *     void vsetStart(vset *set, vsetIterator *it);
+ *     bool vsetNext(vsetIterator *it, void **entryptr);
+ *     void vsetStop(vsetIterator *it);
+ *
+ *-----------------------------------------------------------------------------
+ * Iteration Support
+ *-----------------------------------------------------------------------------
+ *
+ * Iterator structure maintains context across all bucket types:
+ *
+ *     typedef struct vsetIterator {
+ *         raxIterator riter;           // for RAX
+ *         hashtableIterator hiter;     // for HT
+ *         uint32_t viter;              // for VECTOR
+ *         void *vsingle;               // for SINGLE
+ *         vsetBucket *parent_bucket;   // owning bucket
+ *         vsetBucket *bucket;          // active bucket
+ *         void *entry;                 // current entry
+ *         long long bucket_ts;         // for RAX
+ *         int iteration_state;         // internal FSM
+ *     } vsetIterator;
+ * */
+
+#define VOLATILESET_BUCKET_INTERVAL_MAX (1LL << 13LL) // 2^13 = 8192 milliseconds
+#define VOLATILESET_BUCKET_INTERVAL_MIN (1LL << 4LL)  // 2^4 = 16 milliseconds
+
+#define VOLATILESET_VECTOR_BUCKET_MAX_SIZE 127
+
+typedef long long (*vsetGetExpiryFunc)(const void *entry);
+typedef int (*vsetExpiryFunc)(void *entry, void *ctx);
+
+// Generic bucket type
+typedef void vsetBucket;
+
+// vset is just a pointer to a bucket
+typedef vsetBucket *vset;
+
+typedef struct vsetIterator {
+    /* for rax bucket */
+    raxIterator riter;
+    union {
+        /* for hashtable bucket */
+        hashtableIterator hiter;
+        /* for vector bucket */
+        uint32_t viter;
+        /* for single bucket */
+        void *vsingle;
+    };
+    /* the parent of the bucket we are currently iterating on */
+    vsetBucket *parent_bucket;
+    /* the bucket we are currently iterating on */
+    vsetBucket *bucket;
+    /* the pointer entry */
+    void *entry;
+    /* In case of rax encoded set, this is the current iterated bucket timestamp */
+    long long bucket_ts;
+    /* the state of the iteration */
+    int iteration_state;
+} vsetIterator;
+
+bool vsetAddEntry(vset *set, vsetGetExpiryFunc getExpiry, void *entry);
+bool vsetRemoveEntry(vset *set, vsetGetExpiryFunc getExpiry, void *entry);
+bool vsetUpdateEntry(vset *set, vsetGetExpiryFunc getExpiry, void *old_entry, void *new_entry, long long old_expiry, long long new_expiry);
+bool vsetIsEmpty(vset *set);
+void vsetStart(vset *set, vsetIterator *it);
+bool vsetNext(vsetIterator *it, void **entryptr);
+void vsetStop(vsetIterator *it);
+void vsetInit(vset *set);
+void vsetClear(vset *set);
```

Comment:
What's the difference between `vsetInit` and `vsetClear`.

I'm guessing that "init" is really for create.  "clear" seems like emptying the structure.  But I don't see a delete.

Normally I expect something like this:
```
typedef struct vset; // an opaque definition
vset *vsetCreate();  // returns a new vset
void vsetClear(vsetEntryDeleteFn entryDelete);  // empties a vset, calling the callback on each item
void vsetDelete();  // deletes an empty vset
```
Maybe you're looking to be able to have a stack based value?  Where you get a reference to a pointer?

In any case, documentation would help.

Reply:
Clear will indeed empty the vset (do not recall when but this name was suggested at some review point) and init starts a new vset, potentially on the stack as well. I can extend documentation